lib/compiler-singlepass/src/codegen_x64.rs

```diff
@@ -407,25 +407,22 @@ impl<'a> FuncGen<'a> {
     }
 
     fn emit_gas_const(&mut self, cost: u64) {
-        if cost > 0 {
-            // without this, emit_add in emit_gas eliminates the add 0, which leaves CF clobbered
-            if let Ok(cost) = u32::try_from(cost) {
-                self.emit_gas(Location::Imm32(cost));
-            } else {
-                let cost_reg = self.machine.acquire_temp_gpr().unwrap();
-                self.assembler
-                    .emit_mov(Size::S64, Location::Imm64(cost), Location::GPR(cost_reg));
-                self.emit_gas(Location::GPR(cost_reg));
-                self.machine.release_temp_gpr(cost_reg);
-            }
+        if let Ok(cost) = u32::try_from(cost) {
+            self.emit_gas(Location::Imm32(cost));
```

Comment:
This seems somewhat worse now than it was before. Earlier the logic about skipping 0-cost gas instrumentation was local to this function (and duplicated in the callees.) Now you need to reason through “`u32::try_from(0u64)` cannot fail, so for cost=0 we’ll always pass in `Location::Imm32(0)` and that's when it gets skipped…”

But I’m okay with it either way.

Reply:
Hmm so my thinking was, now `emit_gas` works fine with literally any input that does not trigger a panic, and so `emit_gas_const` should not have to worry about it. Does that make sense?

Reply:
It definitely does, I was just making an observation that the optimization for `gas=0` became somewhat less locally reasoned. But as I said its okay either way.